crates/symbolicator/src/server.rs

```diff
@@ -1,12 +1,22 @@
+use std::fs::read;
 use std::net::SocketAddr;
+use std::path::PathBuf;
 
 use anyhow::{Context, Result};
 
+use axum_server::tls_rustls::RustlsConfig;
+use futures::future::try_join_all;
+use futures::future::BoxFuture;
+
 use crate::config::Config;
 use crate::endpoints;
 use crate::services::Service;
 
-/// Starts all actors and HTTP server based on loaded config.
+fn read_pem_file(path: &PathBuf) -> Result<Vec<u8>> {
```

Comment:
I see there is `RustlsConfig::from_pem_file`, so you don’t need to read these files manually.

Reply:
You are wrapping `RustlsConfig::from_pem` is already async, and you are wrapping it in a `block_on`. So I’m not sure what the benefit would be to read the pem files ahead of time with sync code, just to pass them into an async constructor anyway.

Reply:
The main motivation is to be able to provide accurate errors about which file is missing, and to avoid the test-then-open race.

Honestly it's not clear to me why `RustlsConfig::from_pem` is even `async`, it seems to just complicate its use here. But if you would like me to implement this part using `RustlsConfig::from_pem_file`, I'm happy to do so, it's not a big deal.

Reply:
Fair. With `from_pem_file0,` a missing `key.pem` file yields this error and backtrace:
```
  2022-06-13T23:08:36.173987Z  INFO symbolicator::server: Starting HTTP server on 127.0.0.1:3021
    at crates/symbolicator/src/server.rs:62

  2022-06-13T23:08:36.181959Z ERROR symbolicator::logging: failed to start the server

Caused by:
    No such file or directory (os error 2)

Stack backtrace:
   0: backtrace::backtrace::trace_unsynchronized
   1: backtrace::backtrace::trace
   2: anyhow::backtrace::capture::Backtrace::create
   3: anyhow::backtrace::capture::Backtrace::capture
   4: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   5: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
   6: symbolicator::server::run
   7: symbolicator::cli::execute
   8: symbolicator::main
   9: core::ops::function::FnOnce::call_once
  10: std::sys_common::backtrace::__rust_begin_short_backtrace
  11: std::rt::lang_start::{{closure}}
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:259:13
      std::panicking::try::do_call
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:48
      std::panicking::try::do_call
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
      std::panicking::try
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
      std::panic::catch_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
      std::rt::lang_start_internal
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:20
  13: std::rt::lang_start
  14: _main
    at crates/symbolicator/src/logging.rs:85
```
It seems clearer to be able to say `unable to read file: key.pem`.